### ISSUE TITLE
Optimizer Race in Variable Update under PS strategy

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [keras-team/keras#18668]
# Summary
Using tf.optimizers.experimental, variable update step is not sufficiently locked and can race across workers. This can lead to inconsistent updates of model variables/accumulator state. For some models I've observed consistent nan loss/explosions with `tf.optimizers.experimental`, but `tf.optimizers.legacy` training is stable.

There was a pr that partial fixed this [here](https://github.com/keras-team/keras/commit/a32070b2146e8b0442505fb32ab93d8123931364) and also cited an internal google issue. That pr got reverted for no specified reason shortly after [here](https://github.com/keras-team/keras/commit/fb5bf606aba9ac712f0016af3c79d58fac1b5cf5). Manually patching that pr back does eliminate nan/loss explosion issue for some models I tested it with. That pr is mostly improvement, but it does sometimes appear to sometimes cause new error in training. The exact error message after patching that pr back in looks like,

```
self._critical_section.execute(
Node: 'while/critical_section_execute_114/MutexLock'
6 root error(s) found.

(0) UNAVAILABLE:   Stream IDs exhausted
Additional GRPC error information from remote target /job:chief/replica:0/task:0:
:{"created":"@1697949288.512862705","description":"Error received from peer ipv4:10.42.178.57:2222","file":"external/com_github_grpc_grpc/src/core/lib/surface/call.cc","file_line":1056,"grpc_message":"Stream IDs exhausted","grpc_status":14}
 [[{{node while/critical_section_execute_114/MutexLock}}]]
Additional GRPC error information from remote target /job:ps/replica:0/task:5:
:{"created":"@1697949288.513712689","description":"Error received from peer ipv4:10.42.187.169:2222","file":"external/com_github_grpc_grpc/src/core/lib/surface/call.cc","file_line":1056,"grpc_message":"Stream IDs exhausted\nAdditional GRPC error information from remote target /job:chief/replica:0/task:0:\n:{"created":"@1697949288.512862705","description":"Error received from peer ipv4:10.42.178.57:2222","file":"external/com_github_grpc_grpc/src/core/lib/surface/call.cc","file_line":1056,"grpc_message":"Stream IDs exhausted","grpc_status":14}\n\t [[{{node while/critical_section_execute_114/MutexLock}}]]","grpc_status":14}
 [[while/body/_1/while/critical_section_execute_192/ReadVariableOp/_11604]]
```

I'm unsure if this is deadlock related, transient networking issue more likely with criticalsection usage, or something else.

Also while for some models this leads to nan/crash if you check for nans in training, in other models it may lead to silent model quality changing due to undefined ordering behavior of variable updates. While PS is async strategy I'd still except individual optimizer updates to be applied fully (potentially with stale gradients) and be consistent with how tf.optimizers.legacy behaved.  As tf.optimizers.experimental is default for tensorflow >2.11 and with PS strategy may lead to large model quality metric inconsistencies this seems valuable to fix.

One thing I'm confused/unaware is how did legacy optimizer api avoid this race condition issue? Big difference I see in update step is old optimizer tended to use ResourceApplyAdagradV2 vs individual operations in python but unsure that's related to locking difference.

This is fairly model + optimizer specific for how big behavior difference is. Simple model should show inconsistencies in variable update consistency, but metric gap/nan loss is trickier to reproduce. I think model that's somewhat unstable like GAN with separate optimizer for generator and discriminator would be a way to make large gap.

**Comment by sachinprasadhs** [comment]
Hi, Thanks for reporting the issue.
`Keras` is now migrated to `Keras 3` which is a new multi-backend implementation of the Keras API, with support for `TensorFlow, JAX, and PyTorch`.
Could you please test it with the Keras 3 code.

You can install Keras using below steps.

```
pip install -U keras-nightly
import os
os.environ['KERAS_BACKEND'] = 'tensorflow'
import keras
```

**Comment by sachinprasadhs** [comment]
My bad, I did not notice the PS strategy, you can close this issue and follow up the progress in your previous issue here https://github.com/keras-team/keras/issues/18624

**Comment by hmc-cs-mdrissi** [comment]
The two issues are both specific to PS strategy but the bug is very different for the two. This one the bug affects model training stability in any job. Higher amount of workers the more races will happen. The other is specific to multiple gpu devices per worker and crashes immediately. I would expect the fix for each to be separate and mostly unrelated.

**Comment by hmc-cs-mdrissi** [comment]
I'll work on this issue this week. I've been testing a lot of models/comparing legacy.Optimizer vs experimental.Optimizer code and think I have reasonable fix in mind.

The main change will be to add use_locking=True to the Variable update [operations](https://www.tensorflow.org/api_docs/python/tf/Variable#assign). Legacy optimizer had use_locking=True for individual updates which ensured consistency. CriticalSection approach mostly works (40 workers is stable/fine), but at high worker counts may deadlock (80+ workers). use_locking=True is finer lock and with older optimizers have been able to train with pretty high (100+) workers.

With the new keras 3 I expect this will be add a new optional argument [use_locking](https://github.com/keras-team/keras/blob/3028ab52cf2640fea98b4935b8b614540f1587d8/keras/backend/common/variables.py#L120) and only tf backends direct assign will use it. Unsure other backends have equivalent need.
